### PR TITLE
[FIX] CORS error while interacting with any action button on Livechat

### PR DESCRIPTION
**Pull request by murtaza98**
## Proposed changes (including videos or screenshots)

Duplicate of #21406 

PS: Had to close the previous PR because some of the builds were not passing

## Issue(s)

#21033

### CODE REVIEW

File: app/apps/server/communication/uikit.js [hunk 1]
```diff
@@ -14,6 +15,25 @@ const apiServer = express();
 
 apiServer.disable('x-powered-by');
 
+let corsEnabled = false;
+let allowListOrigins = [];
+
+settings.get('API_Enable_CORS', (_, value) => { corsEnabled = value; });
```
Comment (KevLehman) [inline app/apps/server/communication/uikit.js:21]: While testing this, noticed some assets urls are affected by this setting, is that the expected behavior? 🤔

Comment (murtaza98) [reply app/apps/server/communication/uikit.js:21]: Hey @KevLehman Could you share more info about this? I think, ideally it shouldn't affect any such assets url?

File: app/apps/server/communication/uikit.js [hunk 2]
```diff
@@ -14,6 +15,25 @@ const apiServer = express();
...
+	origin: (origin, callback) => {
+		if (!origin || !corsEnabled || (allowListOrigins.includes('*') || allowListOrigins.includes(origin)) || origin === settings.get('Site_Url')) {
+			callback(null, true);
+		} else {
+			callback('Not allowed by CORS', false);
+		}
+	},
+}));
+
 WebApp.connectHandlers.use(apiServer);
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply app/apps/server/communication/uikit.js:37]: Based on the `CORS` package documentation - its not required if we use it at application level which is what we're doing here
image

File: app/apps/server/communication/uikit.js (hunk 1 above)
Comment (murtaza98) [reply app/apps/server/communication/uikit.js:21]: Thanks for pointing this out! It seems this was an issue bcs I was applying the CORS filter on top of app the routers while actually we should just apply it on the `ui.interaction` endpoint. I've fixed the same now :+1:

File: app/apps/server/communication/uikit.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (KevLehman) [reply app/apps/server/communication/uikit.js:37]: Nice 🎉

File: app/apps/server/communication/uikit.js (hunk 1 above)
Comment (KevLehman) [reply app/apps/server/communication/uikit.js:21]: It works as expected now :)

File: app/apps/server/communication/uikit.js [hunk 3]
```diff
@@ -14,6 +15,25 @@ const apiServer = express();
...
+
+settings.get('API_Enable_CORS', (_, value) => { corsEnabled = value; });
+
+settings.get('API_CORS_Origin', (_, value) => {
+	allowListOrigins = value ? value.trim().split(',').map((origin) => String(origin).trim().toLocaleLowerCase()) : [];
+});
+
+const corsOptions = {
+	origin: (origin, callback) => {
+		if (!origin || (corsEnabled && (allowListOrigins.includes('*') || allowListOrigins.includes(origin))) || origin === settings.get('Site_Url')) {
```
Comment (KevLehman) [inline app/apps/server/communication/uikit.js:29]: Be aware there's an open PR #22106 that aims to modify `Site_Url` setting)

File: app/apps/server/communication/uikit.js (hunk 1 above)
Comment (murtaza98) [reply app/apps/server/communication/uikit.js:21]: Great! Could you approve this as well :smile:

**Comment by murtaza98** [comment]
The failing builds is throwing the following error, but I don't think its related to any of the changes we've here :confused: 
image

**Comment by sampaiodiego** [comment]
> The failing builds is throwing the following error, but I don't think its related to any of the changes we've here 
> image

omg.. I don't know why this is happening some times.. one way I came across to solve this is closing the PR and opening a new one based on the same branch.. :(

**Comment by murtaza98** [comment]
> > The failing builds is throwing the following error, but I don't think its related to any of the changes we've here
> > image
> 
> omg.. I don't know why this is happening some times.. one way I came across to solve this is closing the PR and opening a new one based on the same branch.. :(

Okay let me try this

**Pull request closed without merging**